Allow users to run the painless execute API on a remote cluster shard

quux00 opened the pull request:
We added a clusterAlias to the Painless execute Request object, so that index
expressions in the request of the form "myremote:myindex" will be parsed to
set clusterAlias to "myremote" and the index to "myindex".

If clusterAlias is null, then it is executed against a shard on the local cluster, as before.
If clusterAlias is non-null, then the SingleShardTransportAction is sent to the remote
cluster, where it will run the full request (doing remote coordination). 

The clusterAlias field is not part a Writeable field - it not included when serializing
to another node. So when the Action is forwarded to a remote cluster coordinator,
it will just get the index name, and no clusterAlias, indicating that it should run the 
Action locally.

A multi-cluster IT test was added for testing these scenarios.

Closes #92126 and Kibana issue [#145861](https://github.com/elastic/kibana/issues/145861)

quux00 committed:
Allow uses to run the painless execute API on a remote cluster shard.

We added a clusterAlias to the Painless execute Request object, so that index
expressions in the request of the form "myremote:myindex" will be parsed to
set clusterAlias to myremote and the index to "myindex".

If clusterAlias is null, then it is executed against a shard on the local cluster, as before.
If clusterAlias is non-null, then the SingleShardTransportAction is sent to the remote cluster,
where it will run the full request (doing remote coordination). Note that before forwarding
the request to the remote cluster, we nullify the clusterAlias in the request, so that the
remote coordinator will interpret the request to run it locally, rather than try to lookup
a remote cluster.

Closes #92126 and kibana issue #145861

quux00 committed:
testParseClusterAliasAndIndex added - parsing looks good; need more unit tests; all current tests passing

quux00 committed:
Fixed NPE in last commit for cases where context setup is not supplied; new test in PainlessExecuteApiTests

quux00 committed:
Added PainlessExecuteIT test that tests cross-cluster calls

quux00 committed:
Clean up for review

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Updated painless-execute-script end user docs to indicate support for cross-cluster executions

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

javanna commented on docs/painless/painless-guide/painless-execute-script.asciidoc:
```diff
@@ -118,6 +118,11 @@ Document that's temporarily indexed in-memory and accessible from the script.
+alias. For example, `remote1:my_index` indicates that you want to execute
+the painless script against the "my_index" index on the "remote1" cluster. This
+request will be forwarded to the "remote1" cluster if you have
+<<remote-clusters-connect,configured a connection>> to that remote cluster.
```
thanks! I think that we'll need to update this list too: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cross-cluster-search.html#ccs-supported-apis

javanna commented on docs/painless/painless-guide/painless-execute-script.asciidoc:
```diff
@@ -118,6 +118,11 @@ Document that's temporarily indexed in-memory and accessible from the script.
 
 `index`:: (Required, string)
 Index containing a mapping that's compatible with the indexed document.
+You may specific a remote index by prefixing the index with the remote cluster
```
specify?

javanna commented on modules/lang-painless/src/internalClusterTest/java/org/elasticsearch/painless/action/CrossClusterPainlessExecuteIT.java:
```diff
@@ -0,0 +1,197 @@
+            .cluster()
+            .execute(PainlessExecuteAction.INSTANCE, request);
+
+        assertBusy(() -> assertTrue(actionFuture.isDone()));
```
do we need this assertBusy? isn't the following actionGet enough?

quux00 replied:
You are correct. I removed the assertBusy.

javanna commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -203,6 +215,34 @@ static ContextSetup parse(XContentParser parser, Void context) throws IOExceptio
             }
 
+            static Tuple<String, String> parseClusterAliasAndIndex(String indexExpression) {
+                if (indexExpression == null) {
```
should we avoid creating a Tuple if the index is null?

javanna commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -203,6 +215,34 @@ static ContextSetup parse(XContentParser parser, Void context) throws IOExceptio
+                        "Unable to parse one single valid index name from the provided index: [" + indexExpression + "]"
+                    );
+                }
+            }
```
would it make sense to reuse the parsing code from RemoteClusterService somehow?

quux00 replied:
I tried that but it didn't work as I hoped. I'll go back and review it again and document it here.

javanna replied:
I thought so, thanks.

javanna commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -175,6 +181,8 @@ static class ContextSetup implements Writeable, ToXContentObject {
             }
 
+            @Nullable // null means local cluster
+            private final String clusterAlias;  // this field is not Writeable, as it is needed only on the initial receiving node
```
just double checking: the index that was provided is not printed out anywhere in the output, as we only return the result of the script execution?

quux00 replied:
Correct. The output of this endpoint is just:
```
{
  "result": true
}
```
where `true` is the value and can be any object (number, string, bool, etc).

javanna replied:
ok we are good then, i was worried about cases that we have in search, that need to preserve the original index name including the cluster alias to output it as part of the response. Glad we don't have this problem in painless execute.

javanna reviewed (commented):
left a couple of comments, looks like a good direction and thanks for all the tests

quux00 committed:
Fixed references in painless-execute-script.asciidoc. Added painless execute api xref to search-across-clusters.asciidoc

javanna reviewed (approved):
left a couple more comments but LGTM, no need for another round of review

quux00 committed:
Fixed reference to painless-execute-api.html in search-across-clusters.asciidoc

No longer calling parseClusterAliasAndIndex in ContextSetup(StreamInput in) ctor as clusterAlias
will never be present.

quux00 commented on docs/reference/search/search-your-data/search-across-clusters.asciidoc:
```diff
@@ -17,6 +17,7 @@ The following APIs support {ccs}:
 * <<search-template,Search template>>
 * <<multi-search-template,Multi search template>>
 * <<search-field-caps,Field capabilities>>
+* {ref}/painless-execute-api.html[Painless execute API]
```
Fixed.

quux00 committed:
added additional code comments

quux00 commented:
We have marked this with `>bug` since it is a regression - this used to work in a previous version (not sure the exact one) and Kibana has requested we back port this to 8.9.1 so they can have the fix ASAP.
